[SHA1] Add fast path when adding UTF8 bytes from a String that only contains ASCII

cdumez opened the pull request:
#### 484500b2c078b29f80e12b8130c3f95ca698e2d3
<pre>
[SHA1] Add fast path when adding UTF8 bytes from a String that only contains ASCII
<a href="https://bugs.webkit.org/show_bug.cgi?id=274873">https://bugs.webkit.org/show_bug.cgi?id=274873</a>

Reviewed by Darin Adler and Timothy Hatcher.

Add fast path when adding UTF8 bytes from a String that only contains ASCII.
Also fix a bug in our HTTP disk cache hashing function where there was already
a fast path and this fast path was hashing the null terminator, where was the
slow path wasn&apos;t. Align both code paths so that we never hash the null
terminator and bump the HTTP cache version since this change is not backward
compatible.

* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/ParseHash.cpp:
(JSC::ParseHash::ParseHash):
* Source/WTF/wtf/SHA1.cpp:
(WTF::SHA1::addUTF8Bytes):
* Source/WTF/wtf/SHA1.h:
(WTF::SHA1::addUTF8Bytes):
* Source/WebCore/inspector/DOMPatchSupport.cpp:
(WebCore::addStringToSHA1):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::addStringToSHA1):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::hashString):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::computeStringHashForContentBlockerRules):

Canonical link: <a href="https://commits.webkit.org/279513@main">https://commits.webkit.org/279513@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c7d2bfddca02be2f11151123b165af3528ccf8c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6166 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56932 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4377 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55954 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4197 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43468 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2858 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31246 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46382 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24602 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28061 "Found 1 new test failure: imported/w3c/web-platform-tests/html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*) (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3705 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2532 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47012 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58526 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53163 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28813 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3917 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50875 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46549 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50222 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30952 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65467 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7924 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29790 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12475 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
```diff
@@ -109,7 +109,7 @@ class Storage : public ThreadSafeRefCounted<Storage, WTF::DestructionThread::Mai
 
     // Incrementing this number will delete all existing cache content for everyone. Do you really need to do it?
     // FIXME: When this is incremented, remove LegacyCertificateInfoType.
```
Can we do this now?

cdumez replied:
I can check.

darinadler commented on Source/WTF/wtf/SHA1.cpp:
```diff
@@ -203,6 +204,27 @@ void SHA1::reset()
+    if (auto* characters = CFStringGetCStringPtr(string, kCFStringEncodingASCII))
+        addBytes(std::span { byteCast<uint8_t>(characters), static_cast<size_t>(CFStringGetLength(string)) });
+    else
+        addUTF8Bytes(String(string));
```
Can write even more code to avoid allocating a `String` when the `CFStringRef` is all ASCII but `CFStringGetCStringPtr` returns `nullptr`. It would use `CFStringGetBytes`.

cdumez replied:
I'll look into that.

cdumez replied:
I'm thinking of using `CFStringGetCString` which seems simpler to use and is documented as fall back to `CFStringGetCStringPtr`.

cdumez replied:
```
void SHA1::addUTF8Bytes(CFStringRef string)
{
    if (auto* characters = CFStringGetCStringPtr(string, kCFStringEncodingASCII)) {
        addBytes(std::span { byteCast<uint8_t>(characters), static_cast<size_t>(CFStringGetLength(string)) });
        return;
    }

    Vector<char, 1024> buffer(CFStringGetLength(string) + 1);
    if (CFStringGetCString(string, buffer.data(), buffer.size(), kCFStringEncodingASCII)) {
        addBytes(span8(buffer.data()));
        return;
    }

    addUTF8Bytes(String(string));
}
```

cdumez committed:
[SHA1] Add fast path when adding UTF8 bytes from a String that only contains ASCII
https://bugs.webkit.org/show_bug.cgi?id=274873

Reviewed by Darin Adler and Timothy Hatcher.

Add fast path when adding UTF8 bytes from a String that only contains ASCII.
Also fix a bug in our HTTP disk cache hashing function where there was already
a fast path and this fast path was hashing the null terminator, where was the
slow path wasn't. Align both code paths so that we never hash the null
terminator and bump the HTTP cache version since this change is not backward
compatible.

* Source/JavaScriptCore/bytecode/CodeBlockHash.cpp:
(JSC::CodeBlockHash::CodeBlockHash):
* Source/JavaScriptCore/bytecode/ParseHash.cpp:
(JSC::ParseHash::ParseHash):
* Source/WTF/wtf/SHA1.cpp:
(WTF::SHA1::addUTF8Bytes):
* Source/WTF/wtf/SHA1.h:
(WTF::SHA1::addUTF8Bytes):
* Source/WebCore/inspector/DOMPatchSupport.cpp:
(WebCore::addStringToSHA1):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::addStringToSHA1):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::hashString):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::computeStringHashForContentBlockerRules):

Canonical link: https://commits.webkit.org/279513@main
